cEP-0018.md

```diff
@@ -0,0 +1,162 @@
+# Integration of ANTLR into coala core
+
+| Metadata     |                                         |
+| ------------ |-----------------------------------------|
+| cEP          | 0018                                    |
+| Version      | 0.1.0                                   |
+| Title        | Integration of ANTLR into coala core    |
+| Authors      | Viresh Gupta <viresh16118@iiitd.ac.in>  |
+| Status       | Proposed                                |
+| Type         | Feature                                 |
+
+# Abstract
+
+This document describes how an API based on ANTLR will be constructed and
+maintained.
+
+# Introduction
+
+ANTLR provides parsers for various language grammars and thus we can provide
+an interface from within coala and make it available to bear writers so that
+they can write advanced linting bears. This will aim at supporting the more
+flexible visitor based method of AST traversal (as opposed to listener based
+mechanisms).
+
+The proposal is to introduce a parallel concept to the coala-bears library,
+which will be called the coala-ast hereon.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+    1. There will be a separate repository named as coala-ast which will be
+    installable via ```pip install coala-ast```.
+    2. A new interface would be introduced in coala-core (```coala/coalib```),
+    which would be the new ast package. This would be the ```coalib.ast``` package
+    3. The ```coalib.ast``` package will provide endpoints relevant to the visitor
+    model of traversing the ast generated via ANTLR.
+    4. The coala-ast repository will hold parser generated (with python targe)
```

Comment:
The time required to generate a parser on my machine for one single language for a python target isn't much....
```
$ time antlr4 -Dlanguage=Python3 Python3.g4 

real	0m1.184s
user	0m3.328s
sys	0m0.112s
```
But the bigger drawback for making a parser on the fly is we would be mandating the user machine to have a compatible JAVA version installed and working so that the antlr's jar file can be called and executed (though on the CI it won't be a problem I believe)
I couldn't come up with an elegant way to go about this...